design/ceph/osd-migration.md

```diff
@@ -0,0 +1,230 @@
+  5. Provide a new environment variables to the OSD prepare pod running on `OSDReplaceInfo.Node` in case of OSD on disks or `OSDReplaceInfo.Path` in case of OSD on PVCs
+      - `ROOK_REPLACE_OSD_ID`: OSD to be deleted.
+  6. OSD prepare pod job will destroy the OSD based on the environment variables provided above. This process is mentioned in the `Destroy OSD` steps below.
+  7. Continue with the existing flow of creating and upgrading the OSD, with the exception of skipping the upgrade for all the OSDs that are currently present in the `OSDReplaceConfig.OSDs`. Skipping update operation on the OSDs that need to be migrated is important because with OSDs on disks, the OSD prepared pod is updated if there is an update in the spec. This will update the osd store to `bluestore-rdr` in the spec without migrating the OSD. This will cause errors.
+  8. Once the destroy OSD pod is added back, reset `OSDInProgress` to nil. If there is any error during the OSD migration, then don't reset `OSDInProgress`.
+  9. Reconcile the operator and perform as same steps in the next reconcile.
+
+#### Destroy OSD
+
+OSD prepare pod job will perform OSD deletion using following steps.
+- It checks for `ROOK_REPLACE_OSD_ID` arguments.
+- Use `ceph volume list` to fetch the OSD path.
+- Destroy the OSD based on the `ROOK_REPLACE_OSD_ID`
+
+    ```
+    ceph osd destroy {ROOK_REPLACE_OSD_ID} --yes-i-really-mean-it
+    ```
+
+- Once the OSD is destroyed, wipe the backing device. This removes all the data on the device.
+- Once the device is wiped, prepare the OSD with the new store type. Use the same OSD ID to prepare the OSD again. This is done by passing the OSD ID as `--osd-id` flag to `ceph-volume`.
+- Prepare the OSD using the same `ROOK_REPLACE_OSD_ID` and path using following commands:
+
+    ```
+    ceph-volume lvm prepare --osd-id {ROOK_REPLACE_OSD_ID} --data {OSD_PATH}
+    ```
+
+#### Manage OSD migration status
+
+- OSD migration status will managed by `OSDReplaceConfig` API. This will be stored as the config map. For example:
+``` yaml
+apiVersion: v1
+data:
+  config: '{"osdList":{"osds":[{"id":1,"path":"/mnt/set1-data-1qwdpk","node":"minikube-m03"},{"id":2,"path":"/mnt/set1-data-2c8h9k","node":"minikube-m04"}]},"isPgHealthy":true,"OSDInprogress":{"id":0,"path":"/mnt/set1-data-045k6q","node":"minikube-m02"},"startTimeStamp":"2023-07-18T14:41:33.180909077Z"}'
+kind: ConfigMap
+metadata:
+  creationTimestamp: "2023-07-18T14:42:29Z"
+  name: osd-replace-config
+  namespace: rook-ceph
+  ownerReferences:
+  - apiVersion: ceph.rook.io/v1
+    blockOwnerDeletion: true
+    controller: true
+    kind: CephCluster
+    name: rook-ceph
+    uid: e634abc3-39fa-4dd0-a015-753030e64f95
+  resourceVersion: "3580"
+  uid: 7b7d35d4-7807-4286-b185-3f8562cb3cbe
+```
+- This config map will have the `ownerReference` of cephCluster resource.
+
+##### Ceph Cluster Status:
+
+- Presence of this config map `osd-replace-config` will indicate that OSD migration is in progress.
```

Comment:
If the replacement is in progress, maybe the osd reconcile loop can just update the cluster CR status directly, instead of using the configmap to store intermediate status.